pkg/deviceplugin/lister.go

```diff
@@ -59,16 +68,86 @@ func (ml MacvtapLister) Discover(pluginListCh chan dpm.PluginNameList) {
 		plugins = append(plugins, macvtapConfig.Name)
 	}
 
-	pluginListCh <- plugins
+	if len(plugins) > 0 {
+		pluginListCh <- plugins
+	}
+	return config, nil
+}
+
+func discoverByLinks(pluginListCh chan dpm.PluginNameList, nsPath string) {
+	// To know when the manager is stoping, we need to read from pluginListCh.
+	// We avoid reading our own updates by using a middle channel.
+	parentListCh := make(chan []string)
+	defer close(parentListCh)
+
+	sendSuitableParents := func() {
+		var linkNames []string
+		err := ns.WithNetNSPath(nsPath, func(_ ns.NetNS) error {
+			var err error
+			linkNames, err = util.FindSuitableMacvtapParents()
+			return err
+		})
+
+		if err != nil {
+			glog.Errorf("Error while finding links: %v", err)
+		}
+
+		parentListCh <- linkNames
+	}
+
+	// Keep updating on changes for suitable parents, first callback is
+	// guaranteed
+	stop := make(chan struct{})
+	defer close(stop)
+	go util.OnSuitableMacvtapParentEvent(
+		nsPath,
+		sendSuitableParents,
+		stop,
+		func(err error) {
+			glog.Error(err)
+		})
+
+	// Keep forwarding updates to the manager until it closes down
+	for {
+		select {
+		case parentNames := <-parentListCh:
+			pluginListCh <- parentNames
```

Comment:
There is a chance her for a panic because we could attempt to write on a channel that is being closed by the DPM. I dont see any consequence other than an ungraceful shutdown but head up for @phoracek as it it might be interesting to add a stop interface to the lister in the DPM.

Reply:
You made me revisit that ancient code! The DP has `stopWatcher` https://github.com/kubevirt/macvtap-cni/blob/546996e2d3a223166285d0709e52b2760809b987/pkg/deviceplugin/plugin.go#L163. We could listen to it. However, there still may be a race where the `Stop` gets called right between lines 113 and 114 AFAIK. We could add locking to the `Stop` method, but I feel it would make the cost less readable. If we fix it, it should go to a different PR.

Reply:
I think the proper fix should happen on DPM side. Add a separate stop interface to lister, if a lister implements it, then let lister close pluginListCh instead of being closed by DPM.